embassy-imxrt/src/flexcomm/uart.rs

```diff
@@ -0,0 +1,1220 @@
+//! Universal Asynchronous Receiver Transmitter (UART) driver.
+
+use core::future::poll_fn;
+use core::marker::PhantomData;
+use core::sync::atomic::{AtomicU8, Ordering};
+use core::task::Poll;
+
+use embassy_futures::select::{select, Either};
+use embassy_hal_internal::drop::OnDrop;
+use embassy_hal_internal::{Peri, PeripheralType};
+use embassy_sync::waitqueue::AtomicWaker;
+use paste::paste;
+
+use crate::dma::AnyChannel;
+use crate::flexcomm::Clock;
+use crate::gpio::{AnyPin, GpioPin as Pin};
+use crate::interrupt::typelevel::Interrupt;
+use crate::iopctl::{DriveMode, DriveStrength, Inverter, IopctlPin, Pull, SlewRate};
+use crate::pac::usart0::cfg::{Clkpol, Datalen, Loop, Paritysel as Parity, Stoplen, Syncen, Syncmst};
+use crate::pac::usart0::ctl::Cc;
+use crate::sealed::Sealed;
+use crate::{dma, interrupt};
+
+/// Driver move trait.
+#[allow(private_bounds)]
+pub trait Mode: Sealed {}
+
+/// Blocking mode.
+pub struct Blocking;
+impl Sealed for Blocking {}
+impl Mode for Blocking {}
+
+/// Async mode.
+pub struct Async;
+impl Sealed for Async {}
+impl Mode for Async {}
+
+/// Uart driver.
+pub struct Uart<'a, M: Mode> {
+    tx: UartTx<'a, M>,
+    rx: UartRx<'a, M>,
+}
+
+/// Uart TX driver.
+pub struct UartTx<'a, M: Mode> {
+    info: Info,
+    tx_dma: Option<Peri<'a, AnyChannel>>,
+    _phantom: PhantomData<(&'a (), M)>,
+}
+
+/// Uart RX driver.
+pub struct UartRx<'a, M: Mode> {
+    info: Info,
+    rx_dma: Option<Peri<'a, AnyChannel>>,
+    _phantom: PhantomData<(&'a (), M)>,
+}
+
+/// UART config
+#[derive(Clone, Copy)]
+pub struct Config {
+    /// Baudrate of the Uart
+    pub baudrate: u32,
+    /// data length
+    pub data_bits: Datalen,
+    /// Parity
+    pub parity: Parity,
+    /// Stop bits
+    pub stop_bits: Stoplen,
+    /// Polarity of the clock
+    pub clock_polarity: Clkpol,
+    /// Sync/ Async operation selection
+    pub operation: Syncen,
+    /// Sync master/slave mode selection (only applicable in sync mode)
+    pub sync_mode_master_select: Syncmst,
+    /// USART continuous Clock generation enable in synchronous master mode.
+    pub continuous_clock: Cc,
+    /// Normal/ loopback mode
+    pub loopback_mode: Loop,
```

Comment:
Not blocking anything here, but with mspm0 we have a similar parameter.

An open question I have on loopback is whether creating a loopback instance should be possible with no pins, or whether pins should always be demanded.

Reply:
We shouldn't need any pins for internal loopback. It's copying from one FIFO to another. The data never leaves the HW.